pkg/diff/diff.go

```diff
@@ -0,0 +1,691 @@
+		// Don't report error and just continue because older Data type does not include raw file paths.
+		// If dataTypes.Data struct will change, it'll be time to consider unmarhsalling it with interface{}.
+		log.Printf("[WARN] unmarshal of %s failed, probably data type is stale", oldPath)
+		return nil, nil
+	}
+
+	bs, err = extracted.showBlob(extracted.new, newPath)
+	if err != nil {
+		return nil, errors.Wrapf(err, "git show. commit: %s, path: %s", extracted.new, newPath)
+	}
+	var newData dataTypes.Data
+	if err := json.Unmarshal([]byte(bs), &newData); err != nil {
+		log.Printf("[WARN] unmarshal of %s failed, probably data type is stale", newPath)
+		return nil, nil
+	}
+
+	var rawfiles = make(map[string]rawfile)
+	for _, fullpath := range oldData.DataSource.Raws {
+		ss := strings.Split(fullpath, string(filepath.Separator))
+		if len(ss) == 0 {
+			return nil, errors.Errorf("raw path does not contain repository name. path: %s", fullpath)
+		}
+
+		rawfiles[fullpath] = rawfile{
+			repo: ss[0],
+			path: strings.Join(ss[1:], string(filepath.Separator)),
+			old:  true,
+		}
+	}
+	for _, fullpath := range newData.DataSource.Raws {
+		ss := strings.Split(fullpath, string(filepath.Separator))
+		if len(ss) == 0 {
+			return nil, errors.Errorf("raw path does not contain repository name. path: %s", fullpath)
+		}
+
+		f := rawfiles[fullpath]
+		rawfiles[fullpath] = rawfile{
+			repo: ss[0],
+			path: strings.Join(ss[1:], string(filepath.Separator)),
+			old:  f.old,
+			new:  true,
+		}
+	}
+
+	return slices.Collect(maps.Values(rawfiles)), nil
+}
+
+type gitRepo struct {
+	name             string
+	url              string
+	dir              string
+	old              string
+	new              string
+	remotes          map[string]remote
+	filter           string
+	diffAlg          string
+	archiveExhausted bool
+}
+
+type remote struct {
+	name         string
+	url          string
+	archiveIndex int
+	partial      bool
+}
+
+func (g *gitRepo) prepare() error {
+	if _, err := os.Stat(g.dir); err != nil {
+		if !os.IsNotExist(err) {
+			return errors.Wrapf(err, "stat. dir: %s", g.dir)
+		}
+
+		if err := os.MkdirAll(filepath.Dir(g.dir), fs.ModePerm); err != nil {
+			return errors.Wrapf(err, "create directory. dir: %s", filepath.Dir(g.dir))
+		}
+		if err := g.init(); err != nil {
+			return errors.Wrapf(err, "git init")
+		}
+		if err := g.addRemoteAndFetch(0); err != nil {
+			return errors.Wrapf(err, "add remote and fetch")
+		}
+	}
+
+	cmd, stdout, stderr, err := execGitCmd(g.dir, "remote", "--verbose")
+	if err != nil {
+		return errors.Wrapf(err, "git remote. command: %s, stdout: %s, stderr: %s", cmd.String(), stdout, stderr)
+	}
+	for _, s := range strings.Split(stdout, "\n") {
+		ss := strings.Fields(s)
+		if len(ss) == 0 {
+			continue
+		}
+		if len(ss) != 3 {
+			log.Printf("[WARN] unknown remote format. remote: %q", s)
+			continue
+		}
+		if ss[2] != "(push)" {
+			continue
+		}
```

Comment:
It may also be a good idea to fix the length using a switch case.

Reply:
didin't go nicely with switch... leave it as it is 🙏